### PR TITLE
Mxnet reader

### PR DESCRIPTION
**Pull request by Indumathi31**

### CODE REVIEW

File: rocAL/rocAL/source/mxnet_recordio_reader.cpp [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+        _data = new uint8_t[_data_size_to_read];
+        _data_ptr = _data;
+        _file_contents.read((char *)_data_ptr, _data_size_to_read);
+        memcpy(&_magic, _data_ptr, sizeof(_magic));
+        _data_ptr += sizeof(_magic);
+        if(_magic != _kMagic)
+            THROW("ERROR: Invalid MXNet RecordIO: wrong _magic number");
+        memcpy(&_length_flag, _data_ptr, sizeof(_length_flag));
+        _data_ptr += sizeof(_length_flag);
+        _cflag = DecodeFlag(_length_flag);
```
Comment (shobana-mcw) [inline rocAL/rocAL/source/mxnet_recordio_reader.cpp:237]: Is this needed here? 
Does _cflag needs to be a private member variable? 
I could see it being used locally only.

File: rocAL/rocAL/source/mxnet_recordio_reader.cpp [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+        _clength =  DecodeLength(_length_flag);
```
Comment (shobana-mcw) [inline rocAL/rocAL/source/mxnet_recordio_reader.cpp:238]: Same applies for _clength too. Check if he needs to be declared as privare variable.

File: rocAL/rocAL/source/rali_api_data_loaders.cpp [hunk 3]
```diff
@@ -740,6 +740,83 @@ raliJpegCaffeLMDBRecordSourceSingleShard(
...
+        RaliImageColor rali_color_format,
+        unsigned internal_shard_count,
+        bool is_output,
+        bool shuffle,
+        bool loop,
+        RaliImageSizeEvaluationPolicy decode_size_policy,
+        unsigned max_width,
+        unsigned max_height)
+{
+    std::cerr << "\nComes inside raliMXNetRecordSource" << std::endl;
```
Comment (shobana-mcw) [inline rocAL/rocAL/source/rali_api_data_loaders.cpp:756]: Remove the print statement.

File: rocAL/rocAL/source/rali_api_meta_data.cpp [hunk 4]
```diff
@@ -133,6 +133,16 @@ RALI_API_CALL raliCreateTFReaderDetection(RaliContext p_context, const char* sou
     return context->master_graph->create_tf_record_meta_data_reader(source_path , MetaDataReaderType::TF_DETECTION_META_DATA_READER,  MetaDataType::BoundingBox, feature_key_map);
 }
 
+RaliMetaData
```
Comment (shobana-mcw) [inline rocAL/rocAL/source/rali_api_meta_data.cpp:136]: Bring parantheses to next line.

**Review by shobana-mcw — COMMENTED** [review]
Remove the commented lines. 
Check if all variables declared in class is needed. if not change them to local variable.

File: rocAL/rocAL/include/mxnet_recordio_reader.h [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+#include <string>
+#include <memory>
+#include <dirent.h>
+#include <map>
+#include <iterator>
+#include <algorithm>
+#include <fstream>
+#include "reader.h"
+#include <google/protobuf/message_lite.h>
+#include "timing_debug.h"
```
[the remark this answers is not in the document]
Comment (Indumathi31) [reply rocAL/rocAL/include/mxnet_recordio_reader.h:34]: Done

File: rocAL/rocAL/include/mxnet_meta_data_reader.h [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    std::ifstream _file_contents;
+    std::vector<size_t> _index_list;
+    size_t _index, _offset, _file_index;
+    uint8_t* _data, *_data_ptr;
+    const uint32_t _kMagic = 0xced7230a;
+    uint32_t _magic, _length_flag, _cflag, _clength;
+    int64_t _seek_pos, _data_size_to_read;
+    ImageRecordIOHeader _hdr;
+    std::map<std::string, std::shared_ptr<Label>> _map_content;
+    std::map<std::string, std::shared_ptr<Label>>::iterator _itr;
```
[the remark this answers is not in the document]
Comment (Indumathi31) [reply rocAL/rocAL/include/mxnet_meta_data_reader.h:44]: Done

File: rocAL/rocAL/include/mxnet_recordio_reader.h [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+class MXNetRecordIOReader : public Reader{
+public:
+    //! Reads the TFRecord File, and loads the image ids and other necessary info
```
[the remark this answers is not in the document]
Comment (Indumathi31) [reply rocAL/rocAL/include/mxnet_recordio_reader.h:38]: Done

File: rocAL/rocAL/source/mxnet_recordio_reader.cpp [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+        if (_hdr.flag == 0)
+            _image_key = to_string(_hdr.image_id[0]);
+        else
+        {
+            WRN("\nMultiple record reading has not supported");
+            continue;
+        }
+        int64_t data_size = _clength - sizeof(ImageRecordIOHeader);
+        int64_t label_size = _hdr.flag * sizeof(float);
+        int64_t image_size = data_size - label_size;
```
[the remark this answers is not in the document]
Comment (Indumathi31) [reply rocAL/rocAL/source/mxnet_recordio_reader.cpp:250]: Done

File: rocAL/rocAL/source/mxnet_recordio_reader.cpp (hunk 1 above)
Comment (Indumathi31) [reply rocAL/rocAL/source/mxnet_recordio_reader.cpp:237]: Removed it from private

File: rocAL/rocAL/source/mxnet_recordio_reader.cpp (hunk 2 above)
Comment (Indumathi31) [reply rocAL/rocAL/source/mxnet_recordio_reader.cpp:238]: Done

File: rocAL/rocAL/source/rali_api_data_loaders.cpp (hunk 3 above)
Comment (Indumathi31) [reply rocAL/rocAL/source/rali_api_data_loaders.cpp:756]: Done

File: rocAL/rocAL/source/rali_api_meta_data.cpp (hunk 4 above)
Comment (Indumathi31) [reply rocAL/rocAL/source/rali_api_meta_data.cpp:136]: Done

File: rocAL/rocAL/include/mxnet_meta_data_reader.h [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    uint32_t DecodeFlag(uint32_t rec) {return (rec >> 29U) & 7U; };
+    uint32_t DecodeLength(uint32_t rec) {return rec & ((1U << 29U) - 1U); };
+    std::vector<std::tuple<int64_t, int64_t>> _indices; // used to store seek position and record size for a particular record.
+    std::ifstream _file_contents;
+    std::vector<size_t> _index_list;
+    size_t _index, _offset, _file_index;
+    uint8_t* _data, *_data_ptr;
+    const uint32_t _kMagic = 0xced7230a;
+    uint32_t _magic, _length_flag, _cflag, _clength;
+    int64_t _seek_pos, _data_size_to_read;
```
[the remark this answers is not in the document]
Comment (Indumathi31) [reply rocAL/rocAL/include/mxnet_meta_data_reader.h]: Done

File: rocAL/rocAL/source/mxnet_meta_data_reader.cpp [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+
+void MXNetMetaDataReader::release() {
+    _map_content.clear();
+}
+
+void MXNetMetaDataReader::read_images()
+{
+    for(int current_index = 0; current_index < (int)_indices.size(); current_index++ )
+    {
+        const uint32_t _kMagic = 0xced7230a;
```
Comment (Kirthi1212) [inline rocAL/rocAL/source/mxnet_meta_data_reader.cpp:142]: constant values can be in header files

Comment (Katiekat1212) [inline rocAL/rocAL/source/mxnet_meta_data_reader.cpp:142]: Constant can be initialised in header file

File: rocAL/rocAL/source/mxnet_recordio_reader.cpp [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+        _indices.emplace_back(_index_list[i], _index_list[i + 1] - _index_list[i]);
+    read_image_names();
+    return Reader::Status::OK;
+}
+
+size_t MXNetRecordIOReader::get_file_shard_id()
+{
+    if (_batch_count == 0 || _shard_count == 0)
+        THROW("Shard (Batch) size cannot be set to 0")
+    //return (_file_id / (_batch_count)) % _shard_count;
```
Comment (Katiekat1212) [inline rocAL/rocAL/source/mxnet_recordio_reader.cpp:218]: Commented code can be removed

**Pull request closed without merging**